File: lib/utils/observer-filter.ts
/**
 * Utility functions for filtering observer participants
 * Observers should be completely invisible in all UI components
 */

import { Participant } from 'livekit-client';

/**
 * Check if a participant is an observer
 * @param participant The participant to check
 * @returns true if the participant is an observer, false otherwise
 */
export function isObserver(participant: Participant | any): boolean {
  if (!participant) {
    return false;
  }

  // Check metadata for observer type
  try {
    const metadata = JSON.parse(participant.metadata || '{}');
    if (metadata.type === 'observer') {
      return true;
    }
  } catch (error) {
    // Metadata parsing failed, continue to other checks
  }

  // Check identity for observer pattern
  if (participant.identity?.includes('_observer_')) {
    return true;
  }

  return false;
}

/**
 * Filter out observers from a list of participants
 * @param participants Array of participants to filter
 * @returns Array of participants with observers removed
 */
export function filterObservers<T extends Participant | any>(participants: T[]): T[] {
  return participants.filter(p => !isObserver(p));
}

/**
 * Get the count of non-observer participants
 * @param participants Array of participants to count
 * @returns Number of non-observer participants
 */
export function countVisibleParticipants(participants: (Participant | any)[]): number {
  return filterObservers(participants).length;
}

File: almajd_app/lib/utils/observer_filter.dart
import 'package:livekit_client/livekit_client.dart' as lk;

/// Utility functions to filter and detect observer participants
class ObserverFilter {
  /// Check if a participant is an observer
  /// Observers are identified by:
  /// 1. Identity contains '_observer_'
  /// 2. Metadata contains type: 'observer'
  static bool isObserver(dynamic participant) {
    if (participant == null) return false;
    
    String? identity;
    String? metadata;
    
    if (participant is lk.LocalParticipant) {
      identity = participant.identity;
      metadata = participant.metadata;
    } else if (participant is lk.RemoteParticipant) {
      identity = participant.identity;
      metadata = participant.metadata;
    } else {
      return false;
    }
    
    // Check identity for observer pattern
    if (identity != null && identity.contains('_observer_')) {
      return true;
    }
    
    // Check metadata for observer type
    if (metadata != null && metadata.contains('"type":"observer"')) {
      return true;
    }
    
    return false;
  }
  
  /// Filter observers from a list of participants
  static List<T> filterObservers<T>(List<T> participants) {
    return participants.where((participant) => !isObserver(participant)).toList();
  }
  
  /// Count visible (non-observer) participants
  static int countVisibleParticipants(List<dynamic> participants) {
    return participants.where((p) => !isObserver(p)).length;
  }
}

